### PR TITLE
Update Louvain to use new graph primitives and pattern accelerators

### LINKED ISSUE #1220: [ENH] MNMG Louvain code cleanup
PR #1172 created an MNMG Louvain implementation.

In order to achieve this implementation, some temporary APIs were exposed in the graph_view_t object.  These should be cleaned up.

### PR DESCRIPTION
**Pull request by ChuckHastings**
Implement the `update_by_delta_modularity` method using the new graph primitives and pattern accelerators.

This eliminates all of the custom MNMG implementation originally created for MNMG Louvain a few releases ago and replaces it with the new pattern accelerator and graph primitives that have been added in the last couple of releases.

This depends on the following PRs and should not be merged until after them:
* #1394 
* #1399 

closes #1220

### CODE REVIEW

File: cpp/src/experimental/louvain.cuh
Comment (seunghwak) [inline cpp/src/experimental/louvain.cuh:334]: ```copy_v_transform_reduce_key_aggregated_out_nbr``` requires ```map_key_first, map_key_last, map_value_first``` to be pre-shuffled (yes... this might be a bit confusing as relabel doesn't require pre-shuffling, so we may need to make this consistent, but this is the current requirement. ```transform_reduce_by_adj_matrix_row(or col)_key``` returns pre-shuffled key, value pairs, so this helps to avoid unnecessary shuffling.

You need to do something like the following, and pass ```keys.beign(), keys.end(), weights.begin()```

I will submit another PR having fixes in the graph primitives side. After these fixes, I am seeing a failure in thrust::sequence call (https://github.com/rapidsai/cugraph/blob/branch-0.19/cpp/src/experimental/louvain.cuh#L524).

```C++ exception with description: tabulate: failed to synchronize ... ```

```
    if (graph_view_t::is_multi_gpu) {
      rmm::device_uvector<vertex_t> keys(local_num_vertices_, handle_.get_stream());
      thrust::sequence(rmm::exec_policy(stream_)->on(stream_), keys.begin(), keys.end(), base_vertex_id_);
      rmm::device_uvector<weight_t> weights(local_num_vertices_, handle_.get_stream());
      thrust::copy(rmm::exec_policy(stream_)->on(stream_),d_dst_cluster_weights_cache_, d_dst_cluster_weights_cache_ + local_num_vertices_, weights.begin());

      auto const comm_size = handle_.get_comms().get_size();
      rmm::device_uvector<vertex_t> rx_keys(0, handle_.get_stream());
      rmm::device_uvector<weight_t> rx_weights(0, handle_.get_stream());
      auto pair_first = thrust::make_zip_iterator(thrust::make_tuple(keys.begin(), weights.begin()));
      std::forward_as_tuple(std::tie(rx_keys, rx_weights), std::ignore) = groupby_gpuid_and_shuffle_values(handle_.get_comms(), pair_first, pair_first + keys.size(), [key_func = cugraph::experimental::detail::compute_gpu_id_from_vertex_t<vertex_t>{comm_size}]__device__(auto val) { return key_func(thrust::get<0>(val)); }, handle_.get_stream());

      keys = std::move(rx_keys);
      weights = std::move(rx_weights);
    }
```

**Comment by seunghwak** [comment]
Did you merge with the latest cuGraph branch? I am seeing some updates I made in "Files Changed"

File: cpp/src/experimental/louvain.cuh
Comment (ChuckHastings) [reply cpp/src/experimental/louvain.cuh]: Resolved.

**Comment by ChuckHastings** [comment]
> Did you merge with the latest cuGraph branch? I am seeing some updates I made in "Files Changed"

Just double checked.  I did merge them.

I had to make a couple of additional changes from my debugging.  There was a problem in `device_comm.cuh`, I suspect I'm the first code that actually compiles the `device_sendrecv_tuple_iterator_element_impl` code in this way.

I modified `generate_graph_from_edgelist` to allow us to skip renumbering.  1 level of Louvain is deterministic if the graphs are numbered the same.  This allows me to compare single GPU and multi GPU results to validate behavior of Louvain.  If renumbering is on there's no good way to compare results across runs.  The renumbering result is a function of the number of GPUs.

File: cpp/src/experimental/louvain.cuh
Comment (seunghwak) [inline cpp/src/experimental/louvain.cuh:24]: Do you still need this?

File: cpp/src/experimental/louvain.cuh
Comment (seunghwak) [inline cpp/src/experimental/louvain.cuh:69]: What's the purpose of having both ```handle_``` and ```stream_``` as ```stream``` is just ```handle_.get_stream()```? This is like maintaining two copies of the same variable.

File: cpp/src/experimental/louvain.cuh
Comment (seunghwak) [inline cpp/src/experimental/louvain.cuh:60]: And are you copying a handle here?

File: cpp/src/experimental/louvain.cuh
Comment (seunghwak) [inline cpp/src/experimental/louvain.cuh:72]: What's the purpose of creating a separate copy of rank_ here? This will add burden to remember that ```rank_``` and ```handle.get_comms().get_rank()``` are same; maintain those two variables in sync; and this will eat-up additional register/cache line/memory.

File: cpp/src/experimental/louvain.cuh
Comment (seunghwak) [inline cpp/src/experimental/louvain.cuh:264]: Do we really need this function anymore? This may cut one line in the caller site but just add an additional layer of indirection to track.

File: cpp/src/experimental/louvain.cuh
Comment (seunghwak) [inline cpp/src/experimental/louvain.cuh:449]: Better delete temporary print code.

File: cpp/src/experimental/louvain.cuh
Comment (seunghwak) [inline cpp/src/experimental/louvain.cuh:490]: Better delete temporary print code.

File: cpp/tests/community/louvain_mg_test.cpp
Comment (seunghwak) [inline cpp/tests/community/louvain_mg_test.cpp:1]: I recommend renaming this to mg_louvain_test.cpp as the executable name is MG_LOUVAIN_TEST.

File: cpp/tests/community/louvain_mg_test.cpp
Comment (seunghwak) [inline cpp/tests/community/louvain_mg_test.cpp:40]: We're mostly using AnalyticsName_Usecase_t, what's the purpose of introducing a new naming convention?

**Comment by seunghwak** [comment]
> I modified `generate_graph_from_edgelist` to allow us to skip renumbering. 1 level of Louvain is deterministic if the graphs are numbered the same. This allows me to compare single GPU and multi GPU results to validate behavior of Louvain. If renumbering is on there's no good way to compare results across runs. The renumbering result is a function of the number of GPUs.

You mean the outcome is dependent on vertex ordering?

**Comment by ChuckHastings** [comment]
> > I modified `generate_graph_from_edgelist` to allow us to skip renumbering. 1 level of Louvain is deterministic if the graphs are numbered the same. This allows me to compare single GPU and multi GPU results to validate behavior of Louvain. If renumbering is on there's no good way to compare results across runs. The renumbering result is a function of the number of GPUs.
> 
> You mean the outcome is dependent on vertex ordering?

Yes.  The algorithm breaks ties by choosing the vertex with the smaller vertex id making it deterministic with consistent vertex ordering.

File: cpp/src/community/dendrogram.cuh
Comment (ChuckHastings) [reply cpp/src/community/dendrogram.cuh:29]: Fixed

File: cpp/src/community/dendrogram.cuh
Comment (ChuckHastings) [reply cpp/src/community/dendrogram.cuh:34]: The code I copied this from used device_buffer...

Can we type erase and pass a device_uvector back to python?   This is currently C++ only, but I was intending that eventually the python interface would be able to fetch these values and store them in cudf columns.

File: cpp/src/experimental/louvain.cuh
Comment (ChuckHastings) [reply cpp/src/experimental/louvain.cuh:24]: The blank line?  Perhaps not.  It prevents clang-format from reordering includes across blocks, I felt like the patterns should be after the graph and functions definitions... but I haven't verified that it's required.

File: cpp/src/experimental/louvain.cuh
Comment (ChuckHastings) [reply cpp/src/experimental/louvain.cuh:69]: The code was adapted from a previous version before we had handle.  I can replace the stream_ references with handle_.get_stream()

File: cpp/src/experimental/louvain.cuh
Comment (ChuckHastings) [reply cpp/src/experimental/louvain.cuh:60]: handle_ is a const reference, this is just copying the reference.

File: cpp/src/experimental/louvain.cuh
Comment (seunghwak) [reply cpp/src/experimental/louvain.cuh:24]: No,  #include <patterns/copy_v_transform_reduce_in_out_nbr.cuh>

AFAIK, this was used to compute weight_sums which has been replaced by a graph_view_t class member function.

File: cpp/src/experimental/louvain.cuh
Comment (ChuckHastings) [reply cpp/src/experimental/louvain.cuh:264]: Deleted this.  In my original MNMG implementation this was called dozens of times.

File: cpp/src/experimental/louvain.cuh
Comment (ChuckHastings) [reply cpp/src/experimental/louvain.cuh:449]: Thanks.  Hadn't promoted this from draft yet, that was on my todo list.

File: cpp/src/experimental/louvain.cuh
Comment (seunghwak) [reply cpp/src/experimental/louvain.cuh:60]: I see.

File: cpp/src/experimental/louvain.cuh
Comment (ChuckHastings) [reply cpp/src/experimental/louvain.cuh:24]: Ah, yes.  You are correct.  I have deleted that.

Comment (ChuckHastings) [reply cpp/src/experimental/louvain.cuh:24]: Spoke too soon.  I actually still need it for a different purpose.

https://github.com/rapidsai/cugraph/blob/9e4421e743f7ddbbf64a1eefcfa46e3ef4846c53/cpp/src/experimental/louvain.cuh#L316

**Comment by seunghwak** [comment]
> Yes. The algorithm breaks ties by choosing the vertex with the smaller vertex id making it deterministic with consistent vertex ordering.

This sounds against _**graph invariants**_ (https://en.wikipedia.org/wiki/Graph_property) we may try to preserve. Isn't there a better way to break a tie without violating _graph invariants_? (yeah... this sounds impossible; if there are two structurally equivalent communities to move in... no way to distinguish the two without relying on labeling...)

What's the purpose of providing determinism? Just for testing or there is a bigger reason?

File: cpp/src/community/dendrogram.cuh
Comment (seunghwak) [reply cpp/src/community/dendrogram.cuh:34]: @Iroy30 may be able to confirm, but I guess this is possible.

https://github.com/rapidsai/cugraph/blob/branch-0.19/cpp/include/utilities/cython.hpp#L21

rmm::device_uvector is widely used in cython.hpp.

**Comment by ChuckHastings** [comment]
> > Yes. The algorithm breaks ties by choosing the vertex with the smaller vertex id making it deterministic with consistent vertex ordering.
> 
> This sounds against _**graph invariants**_ (https://en.wikipedia.org/wiki/Graph_property) we may try to preserve. Isn't there a better way to break a tie without violating _graph invariants_? (yeah... this sounds impossible; if there are two structurally equivalent communities to move in... no way to distinguish the two without relying on labeling...)
> 
> What's the purpose of providing determinism? Just for testing or there is a bigger reason?

Ease of testing.  Request from cuML for it to be deterministic.

Note that Louvain approximates optimizing modularity, exact solution is NP hard.  The serial implementation of Louvain randomly orders the vertices before each pass.  Consequently it randomly picks one of clusters when it encounters a tie.  The consequence of this is that every time you run Louvain you get a different approximation of the answer.

**Comment by seunghwak** [comment]
> Ease of testing. Request from cuML for it to be deterministic.

Let me think more but I am little worried about the current testing approach as this bypasses all the optimizations relying on renumbering; and leaves those optimizations untested.

The current best approach I can think of is to 1) create a renumbered multi_gpu graph, 2) and using the renumber_map from here, renumber edge list to be used in creating a single_gpu graph, 3) and create a single_gpu graph with renumbering disabled.

The results will be deterministic and SG & MG results will coincide, and we can test optimization disabled path of SG implementation with optimization enabled path of MG implementation, so this will have a better coverage.

**Comment by ChuckHastings** [comment]
> > Ease of testing. Request from cuML for it to be deterministic.
> 
> Let me think more but I am little worried about the current testing approach as this bypasses all the optimizations relying on renumbering; and leaves those optimizations untested.
> 
> The current best approach I can think of is to 1) create a renumbered multi_gpu graph, 2) and using the renumber_map from here, renumber edge list to be used in creating a single_gpu graph, 3) and create a single_gpu graph with renumbering disabled.
> 
> The results will be deterministic and SG & MG results will coincide, and we can test optimization disabled path of SG implementation with optimization enabled path of MG implementation, so this will have a better coverage.

Each renumbering pass in the SG and MG code would diverge.  The results would be deterministic within a dendrogram level.  But as soon as we coarsen the graph, the SG coarsen and the MG coarsen (with num GPUs > 1) will create different graphs and the results will diverge.

I had thought a bit about the fact that we're not testing the renumbering phase.  We can certainly create some tests to specifically test the graph coarsening logic so that we're sure that's tested independently of the Louvain testing.  But I can't come up with a good way to test that.

File: cpp/tests/community/louvain_mg_test.cpp
Comment (ChuckHastings) [reply cpp/tests/community/louvain_mg_test.cpp:40]: Changed Testparams to Usecase

**Comment by seunghwak** [comment]
> > > Ease of testing. Request from cuML for it to be deterministic.
> > 
> > 
> > Let me think more but I am little worried about the current testing approach as this bypasses all the optimizations relying on renumbering; and leaves those optimizations untested.
> > The current best approach I can think of is to 1) create a renumbered multi_gpu graph, 2) and using the renumber_map from here, renumber edge list to be used in creating a single_gpu graph, 3) and create a single_gpu graph with renumbering disabled.
> > The results will be deterministic and SG & MG results will coincide, and we can test optimization disabled path of SG implementation with optimization enabled path of MG implementation, so this will have a better coverage.
> 
> Each renumbering pass in the SG and MG code would diverge. The results would be deterministic within a dendrogram level. But as soon as we coarsen the graph, the SG coarsen and the MG coarsen (with num GPUs > 1) will create different graphs and the results will diverge.
> 
> I had thought a bit about the fact that we're not testing the renumbering phase. We can certainly create some tests to specifically test the graph coarsening logic so that we're sure that's tested independently of the Louvain testing. But I can't come up with a good way to test that.

So, there are two issues.

1) Testing with max_levels = 1
2) Testing when there are more than one levels (so graph coarsening will be called).

In case of 2), neither the proposed approach or the current approach will allow directly comparing SG & MG results.
And we're currently testing only the case 1). In the case of 1), both

1-1) completely skipping renumbering
and
1-2) renumbering edges using the renumber maps from the MG path before creating an SG graph

will work and allow directly comparing SG & MG results.

Are these correct? Assuming this, I think 1-2) has a better coverage than 1-1).

File: cpp/src/experimental/louvain.cuh
Comment (ChuckHastings) [reply cpp/src/experimental/louvain.cuh]: Removed

File: cpp/tests/community/louvain_mg_test.cpp
Comment (ChuckHastings) [reply cpp/tests/community/louvain_mg_test.cpp]: Fixed

File: cpp/src/community/dendrogram.cuh
Comment (ChuckHastings) [reply cpp/src/community/dendrogram.cuh:34]: Looks like I can convert a device_uvector into a device_buffer when it's needed in python:

https://github.com/rapidsai/cugraph/blob/7256f329773afe93deecd43f3cea3eceefc7fed1/cpp/include/utilities/cython.hpp#L200

I'll pursue this change.

**Review by afender — COMMENTED** [review]
Lots of great stuff, I posted some nonblocking comments on API and testing.

**Review by aschaffer — APPROVED** [review]

File: cpp/include/algorithms.hpp
Comment (ChuckHastings) [reply cpp/include/algorithms.hpp:18]: I was hoping to add bindings for a louvain that returns a Dendrogram, eventually.  I had to make the C++ change to support MNMG testing which needed the Dendrogram exposed.

File: cpp/tests/community/mg_louvain_test.cpp
Comment (ChuckHastings) [reply cpp/tests/community/mg_louvain_test.cpp:229]: Comparing the output should work for more than Karate.  I didn't have enough time to verify other test cases and adding RMAT.  I was hoping to expand later.

File: cpp/src/community/louvain.cu
Comment (ChuckHastings) [reply cpp/src/community/louvain.cu:31]: Yes.  The legacy version is still faster on SG than the new version.

File: cpp/include/algorithms.hpp
Comment (ChuckHastings) [reply cpp/include/algorithms.hpp:699]: Current python APIs create columns and send the pointer to the data into the C++ when the data size is known a priori.  I was following that pattern.

File: cpp/include/algorithms.hpp
Comment (ChuckHastings) [reply cpp/include/algorithms.hpp:659]: Removed extraneous text from copy/paste.

File: cpp/include/algorithms.hpp
Comment (afender) [reply cpp/include/algorithms.hpp:699]: Fair point, this made me realize that all the feature I did in the past feature had unknown return size :)

File: cpp/tests/community/mg_louvain_test.cpp
Comment (afender) [reply cpp/tests/community/mg_louvain_test.cpp:229]: Sounds good, yeah this can go in the next PR along with python stuff etc.

File: cpp/include/algorithms.hpp
Comment (afender) [reply cpp/include/algorithms.hpp:18]: I like returning a dendrogram too (probably optionally). 

Did you take a look at SLHC btw? It should return a dendrogram (might be in RAFT already). Better be consistent as we consider adding SLHC to cuGraph also.

File: cpp/src/community/louvain.cu
Comment (seunghwak) [reply cpp/src/community/louvain.cu:31]: So, are we using the legacy graph for this? or are we using the legacy version for SG and the new version for MG?

File: cpp/tests/utilities/generate_graph_from_edgelist.cu
Comment (seunghwak) [inline cpp/tests/utilities/generate_graph_from_edgelist.cu:33]: Do you still need the changes in this file after the test updates?

File: cpp/src/community/louvain.cu
Comment (ChuckHastings) [reply cpp/src/community/louvain.cu:31]: The C++ supports:
* Legacy graph for SG
* new graph for SG
* new graph for MG

The python is currently still using legacy graph for SG and new graph for MG.  I'm doing some performance analysis now of the new graph for SG vice the legacy graph for SG.  It might be necessary to get some optimizations in place in order for the new graph to be fast enough.  My last run today (haven't pushed the latest mods yet, but will shortly) had `web-Google.mtx` running in 14s on the legacy implementation and 56s on the new implementation.  `ljournal-2008.mtx` runs in 250s on legacy and crashes (OOM) on the new implementation.

File: cpp/tests/utilities/generate_graph_from_edgelist.cu
Comment (ChuckHastings) [reply cpp/tests/utilities/generate_graph_from_edgelist.cu:33]: Not necessarily, I will verify this.  It does successfully allow supporting of generating a multi-gpu graph without renumbering - in case we find another need for that.

That can be useful for debugging algorithms (knowing the numbering a priori and that it remains consistent across different shape MG tests).

File: cpp/src/community/louvain.cu
Comment (seunghwak) [reply cpp/src/community/louvain.cu:31]: Oops, ljournal is not large at all and if you see OOM with ljournal, I guess something really shouldn't happen is happening in this case; something we should take a look (maybe not in this PR).

File: cpp/tests/utilities/generate_graph_from_edgelist.cu
Comment (seunghwak) [reply cpp/tests/utilities/generate_graph_from_edgelist.cu:33]: I changed this file significantly in PR https://github.com/rapidsai/cugraph/pull/1443, and this is not necessary for testing, we may better undo this change for now.

File: cpp/src/community/louvain.cu
Comment (seunghwak) [reply cpp/src/community/louvain.cu:31]: I may dig deeper to both performance & memory footprint when I tune primitives for Louvain. The execution time and memory consumption sound very bad.

Comment (seunghwak) [reply cpp/src/community/louvain.cu:31]: We should be able to process 1B-ish edge graph using one GPU. IIRC, ljournal is much smaller than that.

Comment (ChuckHastings) [reply cpp/src/community/louvain.cu:31]: Agreed, that should probably be outside of this PR.  But that's why I've left the legacy implementation in place and the default for python.  When you start tuning these we can evaluate when to delete the legacy implementation.

File: cpp/include/algorithms.hpp
Comment (ChuckHastings) [reply cpp/include/algorithms.hpp:18]: Corey and I talked about the dendrogram and potentially moving it to raft.  We thought waiting until the Louvain work was done in cugraph would be prudent.

File: cpp/tests/utilities/generate_graph_from_edgelist.cu
Comment (ChuckHastings) [reply cpp/tests/utilities/generate_graph_from_edgelist.cu:33]: I will revert tomorrow.

Comment (ChuckHastings) [reply cpp/tests/utilities/generate_graph_from_edgelist.cu]: Done

**Review by seunghwak — APPROVED** [review]
Thanks!

File: cpp/tests/community/mg_louvain_helper.hpp
Comment (rlratzel) [inline cpp/tests/community/mg_louvain_helper.hpp:1]: Are `mg_louvain_helper.[cu|hpp]` really specific to louvain? I'm just wondering if they're generally useful, should they be renamed and placed in the `cugraphtestutil` lib?

File: cpp/tests/community/mg_louvain_test.cpp
Comment (rlratzel) [inline cpp/tests/community/mg_louvain_test.cpp:51]: This is a good `FIXME`, I agree.

**Review by rlratzel — APPROVED** [review]
One minor doc question and a couple of comments, but nothing to hold up my approval now.

File: cpp/tests/community/mg_louvain_helper.hpp
Comment (ChuckHastings) [reply cpp/tests/community/mg_louvain_helper.hpp:1]: They are currently specific to louvain.  I think they might be more generally useful.  But I wanted to have at least one more unit test that used it before moving into a more general location.

File: cpp/include/algorithms.hpp
Comment (ChuckHastings) [reply cpp/include/algorithms.hpp:693]: It's a bit more pervasive.  They are all graph views.  In fact all of our algorithms are using graph views and calling them graphs.

I'm testing a change that changes `louvain` calls to consistently call it a graph_view.  I believe we should do this with all algorithms (I'll try and make this change as I rework things to use the new graph primitives).

File: cpp/tests/community/mg_louvain_helper.hpp
Comment (seunghwak) [reply cpp/tests/community/mg_louvain_helper.hpp:1]: And FYI,

renumber_ext_vertices
https://github.com/rapidsai/cugraph/pull/1443/files#diff-4106792213a4762c10634a44ac8c92f194994ec0b595dba4eccaa822e8353978R215

unrenumber_local_int_vertices
https://github.com/rapidsai/cugraph/pull/1443/files#diff-4106792213a4762c10634a44ac8c92f194994ec0b595dba4eccaa822e8353978R245

and unrenumber_local_int_vertices
https://github.com/rapidsai/cugraph/pull/1443/files#diff-4106792213a4762c10634a44ac8c92f194994ec0b595dba4eccaa822e8353978R279

will be added once PR 1443 gets merged.

There will be some overlaps and some functions here may become obsolete.

**Review by afender — APPROVED** [review]

**Comment by BradReesWork** [comment]
@gpucibot merged

**Comment by BradReesWork** [comment]
@gpucibot merge

**Pull request merged**
